### PR TITLE
FEATURE: asset collection editing

### PR DESCRIPTION
**Pull request by dimaip**
TODO:
- [x] support editing tags belonging to the collection
- [x] give some header to the inspector so it would be visible which kind of entity we are currently editing
- [x] refactor inspector components

**What I did**

Support editing asset collection title in the inspector.

**How to verify it**

Try to edit a title of some asset collection. Don't forget to check optimistic updates.

### CODE REVIEW

File: Classes/GraphQL/Resolver/Type/MutationResolver.php [hunk 1]
```diff
@@ -488,6 +488,34 @@ public function deleteAssetCollection($_, array $variables): array
         ];
     }
 
+    /**
+     * @param $_
+     * @param array $variables
+     * @return AssetCollection|null
+     * @throws Exception
+     */
+    public function updateAssetCollection($_, array $variables): AssetCollection
```
Comment (Sebobo) [inline Classes/GraphQL/Resolver/Type/MutationResolver.php:497]:
```suggestion
    public function updateAssetCollection($_, array $variables): ?AssetCollection
```

File: Resources/Private/GraphQL/schema.root.graphql [hunk 2]
```diff
@@ -84,6 +84,8 @@ type Mutation {
     createAssetCollection(title: String!): AssetCollection!
 
     deleteAssetCollection(id: AssetCollectionId!): DeleteAssetCollectionResult!
+
+    updateAssetCollection(id: AssetCollectionId!, title: String): AssetCollection!
```
Comment (Sebobo) [inline Resources/Private/GraphQL/schema.root.graphql:88]:
```suggestion
    updateAssetCollection(id: AssetCollectionId!, title: String!): AssetCollection!
```

Comment (dimaip) [reply Resources/Private/GraphQL/schema.root.graphql:88]: Title would be optional once we have other properties to update, so I didn't want to force it...

File: Classes/GraphQL/Resolver/Type/MutationResolver.php (hunk 1 above)
Comment (dimaip) [reply Classes/GraphQL/Resolver/Type/MutationResolver.php:497]: What does question mark mean here? Sorry, I'm a graphql noob.

Comment (Sebobo) [reply Classes/GraphQL/Resolver/Type/MutationResolver.php:497]: This is PHP related and just means that `null` is a possible return value. Like you already have in the PHPDoc comment.

File: Resources/Private/GraphQL/schema.root.graphql (hunk 2 above)
Comment (Sebobo) [reply Resources/Private/GraphQL/schema.root.graphql:88]: hm I think it's best to already define everything you want in there and return a `NotImplemented` Exception when it's used without being fully implemented. Then reviews of the API are easier.

Comment (dimaip) [reply Resources/Private/GraphQL/schema.root.graphql]: Dunno, why would we want to transfer extra stuff? E.g. when I only want to set the tags. You had the same logic for editing assets and it kinda made sense...

File: Classes/GraphQL/Resolver/Type/MutationResolver.php (hunk 1 above)
Comment (dimaip) [reply Classes/GraphQL/Resolver/Type/MutationResolver.php:497]: LOL, my brain had thought it was looking at graphql definition 😄

File: Resources/Private/GraphQL/schema.root.graphql (hunk 2 above)
Comment (Sebobo) [reply Resources/Private/GraphQL/schema.root.graphql]: I meant you don't have to make the property required, just fully define what you want to do with the api. Then it makes more sense.

Comment (dimaip) [reply Resources/Private/GraphQL/schema.root.graphql]: What do you mean "fully define what to do with the api"?

Comment (Sebobo) [reply Resources/Private/GraphQL/schema.root.graphql]: I had added a lot of definitions to the Schema in the beginning, even though they were not (fully) implemented yet. That made it easier to review the Schema and get an overview.
You said you want to add other properties to the mutation later. So you could just add them now.

Comment (dimaip) [reply Resources/Private/GraphQL/schema.root.graphql]: ah, gotcha. Well, I already implemented it, see the latest code.

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return (
+        tags.join(',') ===
+        assetCollection.tags
+            .map(tag => tag.label)
+            .sort()
+            .join(',')
+    );
+};
+
+const TagSelectBoxAsset = () => {
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx:20]:
```suggestion
const TagSelectBoxAssetCollection = () => {
```

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    const handleChange = useCallback(
+        newTags => {
+            if (!tagsMatchAssetCollection(newTags, selectedAssetCollection)) {
+                updateAssetCollection({
+                    assetCollection: selectedAssetCollection,
+                    tagNames: newTags
+                })
+                    .then(() => {
+                        Notify.ok(translate('actions.updateAssetCollection.success', 'The asset has been tagged'));
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx:39]: The message id and text should be adapted to what actually happens

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                    })
+                    .catch(({ message }) => {
+                        Notify.error(
+                            translate('actions.updateAssetCollection.error', 'Error while tagging the asset'),
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx:43]: Also here a more precise message and is needed

File: Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx (hunk 3 above)
Comment (Sebobo) [reply Resources/Private/JavaScript/src/components/SideBarRight/Inspector/TagSelectBoxAssetCollection.tsx:20]: Shoudn't it be AssetCollectionTagSelectBox instead?

File: Resources/Private/JavaScript/src/hooks/useSelectTag.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import { selectedAssetCollectionIdState, selectedAssetIdState, selectedTagState } from '../state';
+import selectedInspectorViewState from '../state/selectedInspectorViewState';
+
+const useSelectTag = () => {
+    const setSelectedAssetCollectionId = useSetRecoilState(selectedAssetCollectionIdState);
+    const setSelectedTag = useSetRecoilState(selectedTagState);
+    const setSelectedAssetId = useSetRecoilState(selectedAssetIdState);
+    const setSelectedInspectorView = useSetRecoilState(selectedInspectorViewState);
+
+    return (tag: Tag, assetCollection = null) => {
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useSelectTag.ts:12]: `assetCollection` should have a type specified as it's considered `any` right now

File: Resources/Private/JavaScript/src/hooks/useSelectedAssetCollection.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+import { AssetCollection } from '../interfaces';
+import { selectedAssetCollectionIdState } from '../state';
+import { ASSET_COLLECTION_FRAGMENT } from '../queries';
+
+const useSelectedAssetCollection = (): AssetCollection => {
+    const client = useApolloClient();
+    const selectedAssetCollectionId = useRecoilValue(selectedAssetCollectionIdState);
+
+    // Read asset selection from cache as we can only select asset collections that have been queries before
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useSelectedAssetCollection.ts:12]:
```suggestion
    // Read selection from cache as we can only select asset collections that have been queried before
```

File: Resources/Private/JavaScript/src/hooks/useUpdateAssetCollection.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        action({
+            variables: {
+                id: assetCollection.id,
+                title,
+                tagNames
+            },
+            optimisticResponse: {
+                updateAssetCollection: {
+                    ...assetCollection,
+                    title,
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/hooks/useUpdateAssetCollection.ts:34]: Why is title here twice?

File: Resources/Private/JavaScript/src/state/selectedInspectorViewState.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { atom } from 'recoil';
+
+const selectedInspectorViewState = atom<null | 'asset' | 'assetCollection' | 'tag'>({
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/state/selectedInspectorViewState.ts:3]: Should we introduce a type/constants? Could grow further and we wouldn't need the actual string in other files.

File: Resources/Private/JavaScript/src/components/SideBarRight/CurrentSelection.tsx [hunk 10]
```diff
@@ -31,18 +36,35 @@ const CurrentSelection: React.FC = () => {
...
+            label: selectedAssetCollection.title,
+            icon: 'folder'
+        },
+        selectedAsset && { value: 'asset', label: selectedAsset.label, icon: assetIcon }
+    ].filter(Boolean);
+    const selectedOption = options.find(o => o.value === selectedInspectorView);
+    const value = options.find(o => o.value === selectedInspectorView) ? selectedInspectorView : options[0]?.value;
+
+    // @TODO get rid of this junk code in favour of something like this https://neos-project.slack.com/archives/CUEUD49ED/p1604002816009700
+    React.useEffect(() => {
```
Comment (Sebobo) [inline Resources/Private/JavaScript/src/components/SideBarRight/CurrentSelection.tsx:51]: Just import `useEffect` like `useMemo`

**Pull request merged**